tensorflow_datasets/scripts/refactoring/refactoring_dataset.py

```diff
@@ -0,0 +1,178 @@
+# coding=utf-8
+# Copyright 2020 The TensorFlow Datasets Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+
+from absl import app
+import os
+import posixpath
+import shutil
+import tensorflow as tf
+from tensorflow_datasets.core import naming
+from typing import Dict
+
+"""A refactoring script to aggregate datasets file from diferent directories
+
+my_dataset/
+    fake_data/
+    __init__.py
+    dataset.py
+    dataset_test.py
+    make_fake_data.py
+    checksum.txt
+
+To run this script
+python3 -m tensorflow_datasets.scripts.refactoring.refactoring_dataset
+
+New datasets will be generated in tensorflow_datasets/refactored_dataset
+
+"""
+
+TFDS_DIR = 'tensorflow_datasets'
+NEW_TFDS_DIR = 'tensorflow_datasets/refactored_dataset'
+URL_CHECKSUM_DIR = 'tensorflow_datasets/url_checksums'
+FAKE_DATA_DIR = 'tensorflow_datasets/testing/test_data/fake_examples'
+FAKE_DATA_SCRIPT_DIR = 'tensorflow_datasets/testing/fake_data_generation'
+
+_INIT_FILE = """\
+\"""{dataset_name} dataset.\"""
+from tensorflow_datasets.{dataset_type}.{dataset_name}.{dataset_name} import {dataset_name_cls}
+"""
+
+DATASET_TYPE = ['audio',
+                'image',
+                'image_classification',
+                'object_detection',
+                'question_answering',
+                'structured',
+                'summarization',
+                'text',
+                'translate',
+                'video']
+
+audio_datasets = ['common_voice', 'crema_d', 'dementiabank', 'fuss', 'groove',
```

Comment:
Rather than hardcoding the dataset names, you could auto-detect all registered names with `import tensorflow_datasets as tfds` and `tfds.list_builders()`.

You can then use `builder_cls = tfds.builder_cls('mnist')` to get the class, and `builder_cls.name` for the name and `inspect.getfiles` for the path.

Maybe with:

```py
@dataclasses.dataclass
class BuilderInfo:
  path: pathlib.Path  # Path('.../image/mnist.py')
  type: str  # image, text,...
  name: str  # mnist, ...
  ...


def extract_info(builder_cls: Type[tfds.core.DatasetBuilder]) -> BuilderInfo:
  """Extract path, name,... from the DatasetBuilder class."""
  ...
```

Reply:
Will add this functionality

Reply:
Added `@dataclasses.dataclass`